pkg/clustermanager/cluster_manager.go

```diff
@@ -833,6 +838,21 @@ func (c *ClusterManager) waitForAllControlPlanes(ctx context.Context, cluster *t
 	return nil
 }
 
+func (c *ClusterManager) removeOldWorkerNodeGroups(ctx context.Context, workloadCluster *types.Cluster, provider providers.Provider, currentSpec, newSpec *cluster.Spec) error {
+	workerConfigs := provider.BuildMapForWorkerNodeGroupsByName(newSpec.Spec.WorkerNodeGroupConfigurations)
+	for _, prevWorkerNodeGroupConfig := range currentSpec.Spec.WorkerNodeGroupConfigurations {
+		if _, ok := workerConfigs[prevWorkerNodeGroupConfig.Name]; !ok {
+			workerNodeGroupName := fmt.Sprintf("%s-%s", workloadCluster.Name, prevWorkerNodeGroupConfig.Name)
```

Comment:
This is making an assumption that all providers use the same naming. It's a correct assumption, but I think it goes a little bit against the idea of a provider interface.

What do you think about just returning the objects that need to be deleted from the provider? Like some kind of struct (there might be something already in one of the kubernetes api/runtime packages) that has the type, name and namespace.

That way we don't need to make naming assumptions and we simplify a lot the kubectl code. That keeps cluster manager and kubectl agnostic from the objects created/deleted by the providers.

We could move this method to "select" the worker nodes to delete to a shared package and use it from each provider to get the "to delete" worker nodes and return the associated capi objects.